### ISSUE TITLE
Increased node resource usage under investigation

### ISSUE THREAD

**Issue by zhyatt** [nanocurrency/nano-node#2882]
We have recently observed higher than expected node resource usage on the Nano network. This includes more CPU, disk IO and bandwidth being consumed by elections than is typical compared to the number of transactions being confirmed.

We are currently investigating the root cause and working to resolve these conditions. Areas of investigation include frontier confirmation, sequential voting and elections, related backtracking and how voting and confirm_req are issued for active elections.

For those looking to conduct broad tests on the main network it is recommended those be done on the beta network or temporarily postponed, since the results will be inaccurate.

**Comment by schenkty** [comment]
Here is an excellent example of the issue - https://nanocrawler.cc/explorer/account/nano_1zmm5ywgtmepoxrh3ibqpxjiza1oewr5tkuy54ymuuszzatjwpcikpzy4th4

Open block `D09EA69AB326AFD879A23DA239BB3C77AC71703819A07B7693B22FE921C56A1A` was confirmed within the expected timeframe

The Following Send Block `A05A3DFF184010EADD500AC3DC1D8C3CE8DD6A116D12E2C7B8FF3D0645AA8629` is fully independent at this point as it doesn't need to wait on any previous blocks. However, at the time of writing, it hasn't been confirmed after roughly 30 mins.

**Comment by optout21** [comment]
What I noticed:  after a node restart, after a few minutes, the node stops servicing RPC (or just very slow), and log is full with messages like:

```
[2020-Aug-16 07:28:22.211732]: 
Vote tally for root 3B3C5BD2F617FAEF54BA33E0CFC608B65B361FB54F76941E64B746DD8CEB0CE9
Block 2098335474141DC8AEC5881371C127A121F888C7A8518F0956E1F23FA6548019 weight 1951891762864755856472141278942455113
nano_16d45ow3tsj1y3z9n4satwzxgj6qiue1ggxbwbrj3b33qr58bzchkpsffpx4 0 2098335474141DC8AEC5881371C127A121F888C7A8518F0956E1F23FA6548019
nano_36y9jqqw7jwt4igaef9g8oge3gxa4t7miufz5iaefxy11rd7y3wiz5xfwxnz 0 2098335474141DC8AEC5881371C127A121F888C7A8518F0956E1F23FA6548019
nano_3y77eyoaruwf6x6zij6cn3sixibos31yqtdwo1d8df34ykdnqn9bs31g1acw 0 2098335474141DC8AEC5881371C127A121F888C7A8518F0956E1F23FA6548019
nano_1nanodou3spx43u3e61c6gnd3cj1qdxupgsd6jbhmrcr79py4g3tg94kthh6 0 2098335474141DC8AEC5881371C127A121F888C7A8518F0956E1F23FA6548019
nano_3hrppx3sfxoiycjm9iaqsr3odecgarcxxxhsm41s9pbs75ykambxqhu9ys58 0 2098335474141DC8AEC5881371C127A121F888C7A8518F0956E1F23FA6548019
[2020-Aug-16 07:28:22.216817]: 
Vote tally for root 8DDF4D0A3C9986A2219BE8EED31DCA1003EBD140C278132E3A5959B4B0FBCDC5
Block 8C3666EB9A92BBF04827FC1252A2143745D26A62688373EC887FBD852C7B6065 weight 0
[2020-Aug-16 07:28:22.711306]: 
Vote tally for root 08103335B7D440A03BF5DBE4D4AB5C88B8F9DBBA5A3DF38054B02B327AB2E299
Block CCC1F1CC3D9E3C811FBC827B7C60362BDD553DECF30D21C380A9E128A00D53B1 weight 1830088359621188978523867909160292794
nano_36y9jqqw7jwt4igaef9g8oge3gxa4t7miufz5iaefxy11rd7y3wiz5xfwxnz 0 CCC1F1CC3D9E3C811FBC827B7C60362BDD553DECF30D21C380A9E128A00D53B1
nano_1nanodou3spx43u3e61c6gnd3cj1qdxupgsd6jbhmrcr79py4g3tg94kthh6 0 CCC1F1CC3D9E3C811FBC827B7C60362BDD553DECF30D21C380A9E128A00D53B1
nano_3hrppx3sfxoiycjm9iaqsr3odecgarcxxxhsm41s9pbs75ykambxqhu9ys58 0 CCC1F1CC3D9E3C811FBC827B7C60362BDD553DECF30D21C380A9E128A00D53B1
nano_16d45ow3tsj1y3z9n4satwzxgj6qiue1ggxbwbrj3b33qr58bzchkpsffpx4 0 CCC1F1CC3D9E3C811FBC827B7C60362BDD553DECF30D21C380A9E128A00D53B1
```

**Comment by Laurentiu-Andronache** [comment]
Good luck solving this!

**Comment by udikantz** [comment]
> What I noticed: after a node restart, after a few minutes, the node stops servicing RPC (or just very slow), and log is full with messages like:
> 
> ```
> [2020-Aug-16 07:28:22.211732]: 
> Vote tally for root 3B3C5BD2F617FAEF54BA33E0CFC608B65B361FB54F76941E64B746DD8CEB0CE9
> Block 2098335474141DC8AEC5881371C127A121F888C7A8518F0956E1F23FA6548019 weight 1951891762864755856472141278942455113
> nano_16d45ow3tsj1y3z9n4satwzxgj6qiue1ggxbwbrj3b33qr58bzchkpsffpx4 0 2098335474141DC8AEC5881371C127A121F888C7A8518F0956E1F23FA6548019
> nano_36y9jqqw7jwt4igaef9g8oge3gxa4t7miufz5iaefxy11rd7y3wiz5xfwxnz 0 2098335474141DC8AEC5881371C127A121F888C7A8518F0956E1F23FA6548019
> nano_3y77eyoaruwf6x6zij6cn3sixibos31yqtdwo1d8df34ykdnqn9bs31g1acw 0 2098335474141DC8AEC5881371C127A121F888C7A8518F0956E1F23FA6548019
> nano_1nanodou3spx43u3e61c6gnd3cj1qdxupgsd6jbhmrcr79py4g3tg94kthh6 0 2098335474141DC8AEC5881371C127A121F888C7A8518F0956E1F23FA6548019
> nano_3hrppx3sfxoiycjm9iaqsr3odecgarcxxxhsm41s9pbs75ykambxqhu9ys58 0 2098335474141DC8AEC5881371C127A121F888C7A8518F0956E1F23FA6548019
> [2020-Aug-16 07:28:22.216817]: 
> Vote tally for root 8DDF4D0A3C9986A2219BE8EED31DCA1003EBD140C278132E3A5959B4B0FBCDC5
> Block 8C3666EB9A92BBF04827FC1252A2143745D26A62688373EC887FBD852C7B6065 weight 0
> [2020-Aug-16 07:28:22.711306]: 
> Vote tally for root 08103335B7D440A03BF5DBE4D4AB5C88B8F9DBBA5A3DF38054B02B327AB2E299
> Block CCC1F1CC3D9E3C811FBC827B7C60362BDD553DECF30D21C380A9E128A00D53B1 weight 1830088359621188978523867909160292794
> nano_36y9jqqw7jwt4igaef9g8oge3gxa4t7miufz5iaefxy11rd7y3wiz5xfwxnz 0 CCC1F1CC3D9E3C811FBC827B7C60362BDD553DECF30D21C380A9E128A00D53B1
> nano_1nanodou3spx43u3e61c6gnd3cj1qdxupgsd6jbhmrcr79py4g3tg94kthh6 0 CCC1F1CC3D9E3C811FBC827B7C60362BDD553DECF30D21C380A9E128A00D53B1
> nano_3hrppx3sfxoiycjm9iaqsr3odecgarcxxxhsm41s9pbs75ykambxqhu9ys58 0 CCC1F1CC3D9E3C811FBC827B7C60362BDD553DECF30D21C380A9E128A00D53B1
> nano_16d45ow3tsj1y3z9n4satwzxgj6qiue1ggxbwbrj3b33qr58bzchkpsffpx4 0 CCC1F1CC3D9E3C811FBC827B7C60362BDD553DECF30D21C380A9E128A00D53B1
> ```

I am having the same issue.   the node stops responding to RPC requests. all requests are getting timed out. restart might resolve the issue for a random amount of time

**Comment by zhyatt** [comment]
Thanks for the additional details. We are continuing to investigate both the current situation, which is now impacting the network more broadly than before, and working on a fix for the root cause. At this time we don't have any specific asks for node operators but recommend network usage be minimized where possible.

**Comment by zhyatt** [comment]
At the time this issue was opened the network was generally confirming blocks quickly outside of account `nano_3c1hatdu1rjqo8tpjfnzuwcyaoq5nqmztxhpezeakk7qd9cy7jjstf5nysfp` which had lots of unconfirmed waiting for voting. Later some additional activity degraded network conditions which began impacting the confirmations on accounts across the network. Since then the network had worked through additional blocks and the health is improving, however resource usage is still high and confirmations are not optimal.

We are working on fixes for underlying issues and until that is available we are making the following recommendations to minimize impacts to the network:

- Nodes operators should keep their nodes online as much as possible and avoid unnecessary restarts
- All testing should be moved to the beta network

We are continuing to monitor the network for unusual activity but if anyone sees confirmation times outside of account `nano_3c1hatdu1rjqo8tpjfnzuwcyaoq5nqmztxhpezeakk7qd9cy7jjstf5nysfp` further degrade, or sees other unexpected behavior, please let us know here or on Discord. We will provide further updates if additional action is needed or as progress on the underlying issue is made.

**Comment by clemahieu** [comment]
We are continuing to work on updates for the identified issues and are working on fixes. The account chain with slow confirmations has completed. 

- We are now at the stage in our investigation where we advise that those few nodes remaining behind in cemented block counts to attempt a restart of the node to help catch up.
- The beta network is provided for testing/performance.

We are still seeing what resembles atypical usage on the main network. We are continuing to monitor the network for this but if anyone sees other atypical usage, please let us know here. We will provide further updates if additional action is needed or as additional progress is made.

**Comment by zhyatt** [comment]
https://github.com/nanocurrency/nano-node/releases/tag/V21.2 was released today with updates to address the issues discussed in this ticket. We recommend all node operators upgrade to this build, which also includes optimizations for OpenCL-based work generation. There are no special integration considerations or impacts with this release. Build artifacts can be found in the release linked above or in the docs here: https://docs.nano.org/releases/current-release-notes/.
